inst/scribble/scribble.js

```diff
@@ -0,0 +1,626 @@
+    document.removeEventListener('touchmove', this.eraserCursorMovement)
+
+    document.removeEventListener('keydown', this.undo)
+    document.removeEventListener('keydown', this.redo)
+
+    const drawingCanvas = this.getVisibleSlideCanvasContainers()
+    drawingCanvas.forEach((container) => {
+      container.classList.remove('active', 'erase')
+    })
+
+    this.drawBtn.classList.remove('active')
+    this.eraseBtn.classList.remove('active')
+
+    slideshow.resume()
+  }
+
+  eraserDetectErase (ev) {
+    if (this.mouseDown & this.eraseMode) {
+      const xy = [
+        this.currFabric.getPointer(ev).x,
+        this.currFabric.getPointer(ev).y
+      ]
+      const objs = this.getCanvasPaths()
+      const paths = objs.map((path) => {
+        return path.path
+      })
+
+      const shouldRemove = paths.map((path) => {
+        const toRemove = path.some((point) => {
+          const xcond = Math.abs(point[1] - xy[0]) < this.tolerance
+          const ycond = Math.abs(point[2] - xy[1]) < this.tolerance
+          const both = xcond & ycond
+          return both
+        })
+        return toRemove
+      })
+
+      const pathCache = this.pathCache
+      const currFabric = this.currFabric
+
+      objs.map(function (item, index) {
+        if (shouldRemove[index]) {
+          pathCache.push(item)
+          currFabric.remove(item)
+        }
+      })
+    }
+  }
+
+  removeTransparentEraserPaths () {
+    this.currFabric.forEachObject((obj) => {
+      if (obj.stroke === this.transparent) {
+        this.currFabric.remove(obj)
+      }
+    })
+  }
+
+  initEraserCursor () {
+    const slideArea = document.querySelector('.remark-slides-area')
+    this.eraserCursor = document.createElement('div')
+    this.eraserCursor.classList.add('xe-scribble__cursor__eraser', 'hidden')
+    this.eraserCursor.style.width = this.eraserSize + 'px'
+    this.eraserCursor.style.height = this.eraserSize + 'px'
+
+    slideArea.appendChild(this.eraserCursor)
+  }
+
+  eraserCursorMovement (ev) {
+    var pX
+    var pY
+    if (ev.type === 'mousemove') {
+      pX = ev.pageX
+      pY = ev.pageY
+    } else if (ev.type === 'touchmove') {
+      pX = ev.touches[0].pageX
+      pY = ev.touches[0].pageY
+    }
+    this.eraserCursor.style.top = pY - this.tolerance + 'px'
+    this.eraserCursor.style.left = pX - this.tolerance + 'px'
+  }
+
+  undo (ev) {
+    if (ev.keyCode === 37) {
+      this.removeTransparentEraserPaths()
+
+      if (this.currFabric._objects.length === 0) return
+
+      const removedPath = this.currFabric._objects.pop()
+      this.pathCache.push(removedPath)
+      this.currFabric.renderAll()
+    }
+  }
+
+  redo (ev) {
+    if (ev.keyCode === 39) {
+      this.removeTransparentEraserPaths()
+
+      if (this.pathCache.length === 0) return
+
+      const addedPath = this.pathCache.pop()
+      this.currFabric._objects.push(addedPath)
+      this.currFabric.renderAll()
+    }
+  }
+}
```

Comment:
General comment: Undo/redo just won't be available for touch devices unless there is a keyboard attached. I don't know of a good way around this - there aren't many touch gestures and they are all already used by more important features.

Reply:
As for redo/undo, I don't think we have documented its functionality anywhere as of yet. Maybe this is something that should go in the `xaringanExtra` docs or the demo slide? Just a thought.

Reply:
One other comment, right now the `redo()` function doesn't do anything after a "clear". In other words, line removed by clearing aren't added to the `pathCache` for later adding back. This might be fine/desirable. Let me know what you think.

Reply:
Perhaps undo/redo buttons?  I know it is more on the screen, but they could be small.

@mattwarkentin, what gestures are available?  I've seen apps that use 2-finger tap to undo and 3-finger tap to redo.

I think 3-finger swipe left and right could also work, if it is available.

Reply:
@rpruim Good point. We might be able to use multi-touch tap or swipe gestures. I will wait to see what Garrick thinks and we can go from there.

Reply:
For now, let's stay small. We can potentially add undo/redo buttons to the toolbox if there's enough interest. But for now, I'm okay with the tradeoff of needing to use the eraser on mobile.